src/main/java/com/linecorp/armeria/client/http/HttpRequestSubscriber.java

```diff
@@ -135,8 +139,23 @@ private void writeFirstHeader() {
                          host, firstHeaders.method().name(), firstHeaders.path());
         logBuilder.attr(RequestLog.HTTP_HEADERS).set(firstHeaders);
 
-        encoder.writeHeaders(ctx, id, streamId(), firstHeaders, false).addListener(this);
-        ctx.flush();
+        switch (firstHeaders.method()) {
+            case GET:
+            case HEAD:
+            case PATCH:
+            case DELETE:
+            case TRACE:
+            case CONNECT:
+                // Do not send any content for the methods which is not supposed to have content.
+                // To be precise, the RFC 2616 does not prohibit sending non-empty content for these methods,
+                // but it is discouraged to do so.
+                // See: http://stackoverflow.com/q/978061
```

Comment:
> 1) Write all headers except for content-related ones on header write.
> 2) On the first publishing of data, send the required headers and then send the data.

It is not possible to send two non-informational-class HEADERS frames in an HTTP/2 stream IIRC. Also, not flushing the HEADERS frame until the next element (a DATA frame or a HEADERS frame with endOfStream = true) doesn't sounds like a good idea.

> Also if the problem is that reactive streams API is not rich enough, I would ditch it. Generic APIs don't have much value if they can't solve the required use cases - for armeria it is implementing a compliant HTTP server I think.

Agreed. However, we internally adapt RxJava Observables into a RichPublisher so there might be an unforeseen problem with going this route. How about filing a new issue for this discussion and follow-up there?